packages/peregrine/lib/talons/CheckoutPage/AddressBook/useAddressBook.js

```diff
@@ -0,0 +1,138 @@
+import { useCallback, useEffect, useState, useRef } from 'react';
+import { useMutation, useQuery } from '@apollo/react-hooks';
+
+import { useAppContext } from '../../../context/app';
+import { useCartContext } from '../../../context/cart';
+
+export const useAddressBook = props => {
+    const {
+        mutations: { setCustomerAddressOnCartMutation },
+        queries: { getCustomerAddressesQuery, getCustomerCartAddressQuery },
+        toggleActiveContent
+    } = props;
+
+    const [, { toggleDrawer }] = useAppContext();
+    const [{ cartId }] = useCartContext();
+
+    const addressCount = useRef();
+    const [activeAddress, setActiveAddress] = useState();
+    const [selectedAddress, setSelectedAddress] = useState();
+
+    const [
+        setCustomerAddressOnCart,
+        { loading: setCustomerAddressOnCartLoading }
+    ] = useMutation(setCustomerAddressOnCartMutation);
+
+    const {
+        data: customerAddressesData,
+        error: customerAddressesError,
+        loading: customerAddressesLoading
+    } = useQuery(getCustomerAddressesQuery);
+
+    const {
+        data: customerCartAddressData,
+        error: customerCartAddressError,
+        loading: customerCartAddressLoading
+    } = useQuery(getCustomerCartAddressQuery);
+
+    useEffect(() => {
+        if (customerAddressesError) {
+            console.error(customerAddressesError);
+        }
+
+        if (customerCartAddressError) {
+            console.error(customerCartAddressError);
+        }
+    }, [customerAddressesError, customerCartAddressError]);
+
+    const isLoading =
+        customerAddressesLoading ||
+        customerCartAddressLoading ||
+        setCustomerAddressOnCartLoading;
+
+    const customerAddresses =
+        (customerAddressesData && customerAddressesData.customer.addresses) ||
+        [];
+
+    useEffect(() => {
+        if (customerAddresses.length !== addressCount.current) {
+            // Auto-select newly added address when count changes
+            if (addressCount.current) {
+                const newestAddress =
+                    customerAddresses[customerAddresses.length - 1];
+                setSelectedAddress(newestAddress.id);
+            }
+
+            addressCount.current = customerAddresses.length;
+        }
+    }, [customerAddresses]);
+
+    const handleEditAddress = useCallback(
+        address => {
+            setActiveAddress(address);
+            toggleDrawer('shippingInformation.edit');
+        },
+        [toggleDrawer]
+    );
+
+    const handleAddAddress = useCallback(() => {
+        handleEditAddress();
+    }, [handleEditAddress]);
+
+    const handleSelectAddress = useCallback(addressId => {
+        setSelectedAddress(addressId);
+    }, []);
+
+    if (
```

Comment:
A comment above this condition would be cool - it's a bit terse. I think it auto-selects the cart address in the address book if it finds a match?

Reply:
Comment added in b5aec84.